File: src/com/roomj/simpletitles/SimpleFrameTitleBuilder.java
package com.roomj.simpletitles;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.wm.impl.FrameTitleBuilder;
import org.jetbrains.annotations.NotNull;

public class SimpleFrameTitleBuilder extends FrameTitleBuilder {
  private static FrameTitleBuilder defaultBuilder;

  public static void setDefaultBuilder(FrameTitleBuilder defaultBuilder) {
    SimpleFrameTitleBuilder.defaultBuilder = defaultBuilder;
  }

  @Override
  public String getProjectTitle(@NotNull Project project) {
    return defaultBuilder.getProjectTitle(project);
  }

  @Override
  public String getFileTitle(@NotNull Project project, @NotNull VirtualFile virtualFile) {
    return virtualFile.getName();
  }
}

File: src/main/kotlin/com/roomj/simpletitles/SimpleFrameTitleBuilder.kt
package com.roomj.simpletitles

import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.wm.impl.PlatformFrameTitleBuilder

class SimpleFrameTitleBuilder : PlatformFrameTitleBuilder() {
    private val state: TitleSettingsState = TitleSettingsState.getInstance()

    override fun getProjectTitle(project: Project): String {
        var title = state.projectFormat
        if (title.contains("{DEFAULT}")) {
            title = title.replace("{DEFAULT}", super.getProjectTitle(project))
        }
        if (title.contains("{SIMPLE}")) {
            title = title.replace("{SIMPLE}", project.name)
        }
        return title
    }

    override suspend fun getFileTitleAsync(project: Project, file: VirtualFile): String {
        return customizeTitle(file, super.getFileTitleAsync(project, file))
    }

    override fun getFileTitle(project: Project, file: VirtualFile): String {
        return customizeTitle(file, super.getFileTitle(project, file))
    }

    private fun customizeTitle(file: VirtualFile, original: String): String {
        var title = state.fileFormat
        if (title.contains("{DEFAULT}")) {
            title = title.replace("{DEFAULT}", original)
        }
        if (title.contains("{SIMPLE}")) {
            title = title.replace("{SIMPLE}", file.name)
        }
        return title
    }
}